src/types/redis_tdigest.cc

```diff
@@ -0,0 +1,429 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+
+#include "redis_tdigest.h"
+
+#include <fmt/format.h>
+#include <rocksdb/db.h>
+#include <rocksdb/iterator.h>
+#include <rocksdb/options.h>
+#include <rocksdb/slice.h>
+#include <rocksdb/status.h>
+
+#include <algorithm>
+#include <iterator>
+#include <limits>
+#include <memory>
+#include <range/v3/algorithm/minmax.hpp>
+#include <range/v3/range/conversion.hpp>
+#include <range/v3/view/join.hpp>
+#include <range/v3/view/transform.hpp>
+#include <vector>
+
+#include "db_util.h"
+#include "encoding.h"
+#include "status.h"
+#include "storage/redis_db.h"
+#include "storage/redis_metadata.h"
+#include "types/tdigest.h"
+
+namespace redis {
+
+// It should be replaced by a iteration of the rocksdb iterator
+class DummyCentroids {
+ public:
+  DummyCentroids(const TDigestMetadata& meta_data, std::vector<Centroid> centroids)
+      : meta_data_(meta_data), centroids_(std::move(centroids)) {}
+  class Iterator {
+   public:
+    Iterator(std::vector<Centroid>::const_iterator&& iter, const std::vector<Centroid>& centroids)
+        : iter_(iter), centroids_(centroids) {}
+    std::unique_ptr<Iterator> Clone() const {
+      if (iter_ != centroids_.cend()) {
+        return std::make_unique<Iterator>(std::next(centroids_.cbegin(), std::distance(centroids_.cbegin(), iter_)),
+                                          centroids_);
+      }
+      return std::make_unique<Iterator>(centroids_.cend(), centroids_);
+    }
+    bool Next() {
+      if (Valid()) {
+        std::advance(iter_, 1);
+      }
+      return iter_ != centroids_.cend();
+    }
+    bool Prev() {
+      if (Valid()) {
+        if (iter_ != centroids_.cbegin()) {
+          std::advance(iter_, -1);
+        }
+      }
+      return Valid();
+    }
+    bool Valid() const { return iter_ != centroids_.cend(); }
+    StatusOr<Centroid> GetCentroid() const {
+      if (iter_ == centroids_.cend()) {
+        return {::Status::NotOK, "invalid iterator during decoding tdigest centroid"};
+      }
+      return *iter_;
+    }
+
+   private:
+    std::vector<Centroid>::const_iterator iter_;
+    const std::vector<Centroid>& centroids_;
+  };
+
+  std::unique_ptr<Iterator> Begin() { return std::make_unique<Iterator>(centroids_.cbegin(), centroids_); }
+  std::unique_ptr<Iterator> End() { return std::make_unique<Iterator>(centroids_.cend(), centroids_); }
+  double TotalWeight() const { return static_cast<double>(meta_data_.total_weight); }
+  double Min() const { return meta_data_.minimum; }
+  double Max() const { return meta_data_.maximum; }
+  uint64_t Size() const { return meta_data_.merged_nodes; }
+
+ private:
+  const TDigestMetadata& meta_data_;
+  std::vector<Centroid> centroids_;
+};
+
+uint64_t constexpr kMaxBufferSize = 1 * 1024;  // 1k doubles
```

Comment:
1k doubles means a 8KB buffer size? Or should we rename this to `kMaxElements` or something else?

Reply:
Hi @mapleFU ,

Got it! Thanks for your suggestion! 😄

I have renamed the variable.

Best Regards,
Edward